tests/PySys/environments/environment_c8y.py

```diff
@@ -10,7 +11,63 @@
 """
 
 
+class Cumulocity(object):
+
+    c8y_url = ""
+    tenant_id = ""
+    username = ""
+    password = ""
+    auth = ""
+
+    def __init__(self, c8y_url, tenant_id, username, password):
+        self.c8y_url = c8y_url
+        self.tenant_id = tenant_id
+        self.username = username
+        self.password = password
+
+        self.auth = ('%s/%s' % (self.tenant_id, self.username), self.password)
+
+    def request(self, method, url_path, **kwargs) -> requests.Response:
+        return requests.request(method, self.c8y_url + url_path, auth=self.auth, **kwargs)
+
+    def getAllDevices(self) -> requests.Response:
+        params = {
+            "fragmentType": "c8y_IsDevice"
+        }
+        res = requests.get(
+            url=self.c8y_url + "/inventory/managedObjects", params=params, auth=self.auth)
+
+        return self.toJsonResponse(res)
+
+    def toJsonResponse(self, res: requests.Response):
+        if res.status_code != 200:
+            raise Exception(
+                "Received invalid response with exit code: {}, reason: {}".format(res.status_code, res.reason))
+        return json.loads(res.text)
+
+    def getAllDevicesByType(self, type: str) -> requests.Response:
+        params = {
+            "fragmentType": "c8y_IsDevice",
+            "type": type,
+        }
+        res = requests.get(
+            url=self.c8y_url + "/inventory/managedObjects", params=params, auth=self.auth)
+        return self.toJsonResponse(res)
+
+    def getAllThinEdgeDevices(self) -> requests.Response:
+        return self.getAllDevicesByType("thin-edge.io")
+
+    def getThinEdgeDeviceByName(self, device_id: str):
+        json_response = self.getAllDevicesByType("thin-edge.io")
+        for device in json_response['managedObjects']:
+            if device_id in device['name']:
+                return device
+        return None
+
+
```

Comment:
Isn't the above code is specific to this test, why to put it in environment, I feel better to put it in run.py, if its not reused by other tests.

Reply:
The idea was to create a common Cumulocity abstraction for all Cumulocity REST calls like get_device_type, get_pending_operations etc which can be reused across tests. It definitely shouldn't be limited to a test alone as I'm sure more tests in future will need to query devices and do other operations as well on the Cumulocity device twin. I put it here because this was the environment that's used by tests connecting to Cumulocity and those tests are more likely to do Cumulocity queries as well.

Reply:
Don't we already have c8y environment to connect to c8y?. I still feel that better to move these functions to test-specific files, till these are shared by other tests.

Reply:
I see this as a good opportunity to refactor as the code grows. I also would like to combine more common functionality. In the long run, I see a python package that we just can import.
Maybe even somebody has created something similar.
BTW: We just found this here https://github.com/SilverLaius/cumulocitypython maybe interresting.

Reply:
@PradeepKiruvale As discussed, this newly added class is not one that must be extended by tests connecting to C8y but rather one that's made available to all those tests extended from the existing `EnvironmentC8y` class to make any Cumulocity REST calls.